File: src/main/java/pw/mihou/nexus/core/managers/core/NexusCommandManagerCore.java
package pw.mihou.nexus.core.managers.core;

import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.ApplicationCommand;
import org.javacord.api.interaction.SlashCommand;
import org.javacord.api.interaction.SlashCommandInteraction;
import org.javacord.api.util.logging.ExceptionLogger;
import pw.mihou.nexus.Nexus;
import pw.mihou.nexus.core.NexusCore;
import pw.mihou.nexus.core.logger.adapters.NexusLoggingAdapter;
import pw.mihou.nexus.core.managers.facade.NexusCommandManager;
import pw.mihou.nexus.core.managers.records.NexusCommandIndex;
import pw.mihou.nexus.features.command.core.NexusCommandCore;
import pw.mihou.nexus.features.command.facade.NexusCommand;

import java.util.*;

public class NexusCommandManagerCore implements NexusCommandManager {

    private final Map<String, NexusCommand> commands = new HashMap<>();
    private final Map<Long, String> indexes = new HashMap<>();
    private final NexusCore nexusCore;
    private static final NexusLoggingAdapter logger = NexusCore.logger;

    /**
     * Creates a new Nexus Command Manager that is utilized to manage commands,
     * index commands, etc.
     *
     * @param nexusCore The nexus core that is in charge of this command manager.
     */
    public NexusCommandManagerCore(NexusCore nexusCore) {
        this.nexusCore = nexusCore;
    }

    @Override
    public Collection<NexusCommand> getCommands() {
        return commands.values();
    }

    @Override
    public Nexus addCommand(NexusCommand command) {
        commands.put(((NexusCommandCore) command).uuid, command);

        return nexusCore;
    }

    @Override
    public Optional<NexusCommand> getCommandById(long id) {
        return Optional.ofNullable(commands.get(indexes.getOrDefault(id, null)));
    }

    @Override
    public Optional<NexusCommand> getCommandByUUID(String uuid) {
        return Optional.ofNullable(commands.get(uuid));
    }

    @Override
    public Optional<NexusCommand> getCommandByName(String name, long server) {
        return getCommands().stream()
                .filter(nexusCommand ->
                        nexusCommand.getName().equalsIgnoreCase(name) && nexusCommand.getServerIds().contains(server)
                )
                .findFirst();
    }

    @Override
    public List<NexusCommandIndex> export() {
        return indexes.entrySet()
                .stream()
                .map(entry -> new NexusCommandIndex(commands.get(entry.getValue()), entry.getKey()))
                .toList();
    }

    /**
     * This performs indexing based on the data analyzed from the
     * {@link SlashCommandCreateEvent} and returns the results for post-processing
     * from the {@link NexusCore}. This is what we call dynamic indexing.
     *
     * @param event The event to handle.
     */
    public Optional<NexusCommand> acceptEvent(SlashCommandCreateEvent event) {
        SlashCommandInteraction interaction = event.getSlashCommandInteraction();

        if (getCommandById(interaction.getCommandId()).isPresent()) {
            return getCommandById(interaction.getCommandId());
        }

        if (interaction.getServer().isPresent()) {
            return getCommands().stream()
                    .filter(nexusCommand -> nexusCommand.getName().equalsIgnoreCase(interaction.getCommandName())
                            && nexusCommand.getServerIds().contains(interaction.getServer().get().getId()))
                    .findFirst()
                    .or(() -> getCommandByName(interaction.getCommandName()))
                    .map(command -> {
                        indexes.put(interaction.getCommandId(), ((NexusCommandCore) command).uuid);
                        return command;
                    });
        }

        return getCommandByName(interaction.getCommandName()).map(command -> {
            indexes.put(interaction.getCommandId(), ((NexusCommandCore) command).uuid);

            return command;
        });
    }

    @Override
    public void index() {
        logger.info("Nexus is now performing command indexing, this will delay your boot time by a few seconds but improve performance and precision in look-ups...");
        long start = System.currentTimeMillis();
        nexusCore.getEngineX().awaitAvailable().thenAcceptAsync(shard -> {
           Set<SlashCommand> slashCommands = shard.getGlobalSlashCommands().join();

           // Clearing the entire in-memory index mapping to make sure that we don't have any outdated indexes.
           indexes.clear();

           for (SlashCommand slashCommand : slashCommands) {
               index(slashCommand);
           }

           Set<Long> servers = new HashSet<>();
           for (NexusCommand serverCommand : getServerCommands()) {
               servers.addAll(serverCommand.getServerIds());
           }

           for (long server : servers) {
               if (server == 0L) continue;

               Set<SlashCommand> $slashCommands = nexusCore.getEngineX().await(server)
                       .thenComposeAsync(Server::getSlashCommands).join();

               for (SlashCommand slashCommand : $slashCommands) {
                   index(slashCommand);
               }
           }

            logger.info("All global and server slash commands are now indexed. It took {} milliseconds to complete indexing.", System.currentTimeMillis() - start);
        }).exceptionally(ExceptionLogger.get()).join();
    }

    @Override
    public void index(NexusCommand command, long snowflake) {
        indexes.put(snowflake, ((NexusCommandCore) command).uuid);
    }

    @Override
    public void index(ApplicationCommand applicationCommand) {
        long serverId = applicationCommand.getServerId().orElse(-1L);

        if (serverId == -1L) {
            for (NexusCommand command : commands.values()) {
                if (!command.getServerIds().isEmpty()) continue;
                if (!command.getName().equalsIgnoreCase(applicationCommand.getName())) continue;

                index(command, applicationCommand.getApplicationId());
                break;
            }
            return;
        }

        for (NexusCommand command : commands.values()) {
            if (command.getServerIds().contains(serverId)) continue;
            if (!command.getName().equalsIgnoreCase(applicationCommand.getName())) continue;

            index(command, applicationCommand.getApplicationId());
            break;
        }
    }

}

File: src/main/java/pw/mihou/nexus/core/managers/core/NexusCommandManagerCore.kt
package pw.mihou.nexus.core.managers.core

import org.javacord.api.event.interaction.MessageContextMenuCommandEvent
import org.javacord.api.event.interaction.SlashCommandCreateEvent
import org.javacord.api.event.interaction.UserContextMenuCommandEvent
import org.javacord.api.interaction.ApplicationCommand
import org.javacord.api.interaction.ApplicationCommandInteraction
import org.javacord.api.util.logging.ExceptionLogger
import pw.mihou.nexus.Nexus
import pw.mihou.nexus.configuration.modules.info
import pw.mihou.nexus.features.commons.NexusApplicationCommand
import pw.mihou.nexus.core.managers.facade.NexusCommandManager
import pw.mihou.nexus.core.managers.indexes.IndexStore
import pw.mihou.nexus.core.managers.indexes.defaults.InMemoryIndexStore
import pw.mihou.nexus.core.managers.indexes.exceptions.IndexIdentifierConflictException
import pw.mihou.nexus.core.managers.records.NexusMetaIndex
import pw.mihou.nexus.features.command.core.NexusCommandCore
import pw.mihou.nexus.features.command.facade.NexusCommand
import pw.mihou.nexus.features.contexts.enums.ContextMenuKinds
import pw.mihou.nexus.features.contexts.NexusContextMenu
import kotlin.collections.HashMap
import kotlin.collections.HashSet

class NexusCommandManagerCore internal constructor() : NexusCommandManager  {
    private val commandsDelegate: MutableMap<String, NexusCommand> = HashMap()
    private val contextMenusDelegate: MutableMap<String, NexusContextMenu> = HashMap()

    override val contextMenus: Collection<NexusContextMenu>
        get() = contextMenusDelegate.values

    override val commands: Collection<NexusCommand>
        get() = commandsDelegate.values

    override var indexStore: IndexStore = InMemoryIndexStore()

    override fun add(command: NexusCommand): NexusCommandManager {
        if (commandsDelegate[command.uuid] != null)
            throw IndexIdentifierConflictException(command.name)

        commandsDelegate[(command as NexusCommandCore).uuid] = command
        return this
    }

    override fun add(contextMenu: NexusContextMenu): NexusCommandManager {
        if (contextMenusDelegate[contextMenu.uuid] != null)
            throw IndexIdentifierConflictException(contextMenu.name)

        contextMenusDelegate[contextMenu.uuid] = contextMenu
        return this
    }

    override operator fun get(applicationId: Long): NexusCommand? = indexStore[applicationId]?.takeCommand()
    override operator fun get(uuid: String): NexusCommand? = commandsDelegate[uuid]
    override operator fun get(name: String, server: Long?): NexusCommand? {
        return commands.firstOrNull { command ->
            when (server) {
                null -> {
                    command.name.equals(name, ignoreCase = true)
                }
                else -> {
                    command.name.equals(name, ignoreCase = true) && command.serverIds.contains(server)
                }
            }
        }
    }

    override fun getContextMenu(applicationId: Long): NexusContextMenu? {
        return indexStore[applicationId]?.takeContextMenu()
    }

    override fun getContextMenu(uuid: String): NexusContextMenu? {
        return contextMenusDelegate[uuid]
    }

    override fun getContextMenu(name: String, kind: ContextMenuKinds, server: Long?): NexusContextMenu? {
        return contextMenus.firstOrNull { contextMenu ->
            when(server) {
                null -> {
                    contextMenu.name.equals(name, ignoreCase = true) && contextMenu.kind == kind
                }
                else -> {
                    contextMenu.name.equals(name, ignoreCase = true) && contextMenu.serverIds.contains(server) && contextMenu.kind == kind
                }
            }
        }
    }

    override fun export(): List<NexusMetaIndex> {
        return indexStore.all()
    }

    private fun toIndex(applicationCommandId: Long, command: String, server: Long?): NexusMetaIndex {
        return NexusMetaIndex(command = command, applicationCommandId = applicationCommandId, server = server)
    }

    fun acceptEvent(event: SlashCommandCreateEvent): NexusCommand? {
        val interaction = event.slashCommandInteraction

        val indexedCommand = get(interaction.commandId)
        if (indexedCommand != null) {
            return indexedCommand
        }

        return if (interaction.server.isPresent) {
            val server = interaction.server.get().id

            return when (val command = get(interaction.commandName, server)) {
                null -> index(interaction, get(interaction.commandName, null), null)
                else -> index(interaction, command, server)
            }

        } else index(interaction, get(interaction.commandName, null), null)
    }

    fun acceptEvent(event: UserContextMenuCommandEvent): NexusContextMenu? {
        return acceptContextMenuEvent(ContextMenuKinds.USER, event.userContextMenuInteraction)
    }

    fun acceptEvent(event: MessageContextMenuCommandEvent): NexusContextMenu? {
        return acceptContextMenuEvent(ContextMenuKinds.MESSAGE, event.messageContextMenuInteraction)
    }

    private fun acceptContextMenuEvent(kind: ContextMenuKinds, interaction: ApplicationCommandInteraction): NexusContextMenu? {
        val indexedContextMenu  = getContextMenu(interaction.commandId)
        if (indexedContextMenu != null) {
            return indexedContextMenu
        }

        return if (interaction.server.isPresent) {
            val server = interaction.server.get().id

            return when (val contextMenu = getContextMenu(interaction.commandName, kind, server)) {
                null -> index(interaction, getContextMenu(interaction.commandName, kind, null), null)
                else -> index(interaction, contextMenu, server)
            }
        } else index(interaction, getContextMenu(interaction.commandName, kind, null), null)
    }

    private fun <Command: NexusApplicationCommand> index(interaction: ApplicationCommandInteraction, uuid: Command?, server: Long?): Command? {
        return uuid?.apply { indexStore.add(toIndex(interaction.commandId, this.uuid, server)) }
    }

    override fun index() {
        Nexus.configuration.loggingTemplates.INDEXING_COMMANDS.info()

        val start = System.currentTimeMillis()
        Nexus.express
            .awaitAvailable()
            .thenAcceptAsync { shard ->
                val applicationCommands  = shard.globalApplicationCommands.join()
                indexStore.clear()
                for (applicationCommand in applicationCommands) {
                    index(applicationCommand)
                }

                val servers: MutableSet<Long> = HashSet()
                for (serverCommand in serverCommands) {
                    servers.addAll(serverCommand.serverIds)
                }
                for (serverContextMenu in serverContextMenus) {
                    servers.addAll(serverContextMenu.serverIds)
                }

                for (server in servers) {
                    if (server == 0L) continue
                    val applicationCommandSet: Set<ApplicationCommand>  = Nexus.express
                        .await(server)
                        .thenComposeAsync { it.api.getServerApplicationCommands(it) }
                        .join()

                    for (applicationCommand in applicationCommandSet) {
                        index(applicationCommand)
                    }
                }

                Nexus.configuration.loggingTemplates.COMMANDS_INDXED(System.currentTimeMillis() - start).info()
            }
            .exceptionally(ExceptionLogger.get())
            .join()
    }

    private fun <Command: NexusApplicationCommand> manifest(command: Command, snowflake: Long, server: Long?) =
        toIndex(applicationCommandId = snowflake, command = command.uuid, server = server)

    override fun <Command: NexusApplicationCommand> index(command: Command, snowflake: Long, server: Long?) {
        indexStore.add(toIndex(applicationCommandId = snowflake, command = command.uuid, server = server))
    }

    override fun index(applicationCommandList: Set<ApplicationCommand>) {
        val indexes = mutableListOf<NexusMetaIndex>()
        for (applicationCommand in applicationCommandList) {
            val serverId: Long? = applicationCommand.serverId.orElse(null)

            if (serverId == null) {
                for (command in commands) {
                    if (command.serverIds.isNotEmpty()) continue
                    if (!command.name.equals(applicationCommand.name, ignoreCase = true)) continue

                    indexes.add(manifest(command, applicationCommand.id, null))
                    break
                }
            } else {
                for (command in commands) {
                    if (!command.name.equals(applicationCommand.name, ignoreCase = true)) continue
                    if (!command.serverIds.contains(serverId)) continue

                    indexes.add(manifest(command, applicationCommand.id, serverId))
                    break
                }
            }
        }

        indexStore.addAll(indexes)
    }

    override fun index(applicationCommand: ApplicationCommand) {
        val serverId: Long? = applicationCommand.serverId.orElse(null)

        if (serverId == null) {
            for (command in commands) {
                if (command.serverIds.isNotEmpty()) continue
                if (!command.name.equals(applicationCommand.name, ignoreCase = true)) continue

                index(command, applicationCommand.id, null)
                break
            }
            return
        } else {
            for (command in commands) {
                if (!command.name.equals(applicationCommand.name, ignoreCase = true)) continue
                if (!command.serverIds.contains(serverId)) continue

                index(command, applicationCommand.id, serverId)
                break
            }
        }
    }
}